spec/models/grace_period_submission_rule_spec.rb

```diff
@@ -5,6 +5,28 @@
     expect(rule.save).to be_truthy
   end
 
+  context 'When the student did not submit any assignment files' do
+    describe '#apply_submission_rule' do
+      before :each do
+        @group = create(:group)
+        @grouping = create(:grouping, group: @group)
+        @membership = create(:student_membership,
+                             grouping: @grouping,
+                             membership_status: StudentMembership::STATUSES[:inviter])
+        @assignment = @grouping.assignment
+        @rule = GracePeriodSubmissionRule.new
+        @assignment.replace_submission_rule(@rule)
+        GracePeriodDeduction.destroy_all
+        @submission = create(:submission, revision_timestamp: nil)
+        @rule.save
+      end
+      it 'should not apply any submission rule' do
```

Comment:
I think this is a good start but there are more important things you should be testing here as well. Think about what applying each submission rule actually does (to the students, the results, the marks, etc.) and write some tests to check those as well.